lib/testing/matrix.js

```diff
@@ -0,0 +1,185 @@
+    *
+    * @returns {string} JSON format of parameters
+    */
+    serialize() {
+        return this.paramsToString;
+    }
+
+    /**
+    * Generate the matrix and call all callback enter on @generate and
+    * if methods. Must be run as the father matrix for running all the test
+    *
+    * @returns {undefined}
+    */
+    execute() {
+        /**
+        * Check if @generate method has been already called
+        */
+        if (this.callback) {
+            /**
+            * Generate matrix element by element, we call the @execute
+            * method until that we specialize all element send
+            * by generate method
+            */
+            if (this.elementsToSpecialize.length !== 0) {
+                /**
+                * Make copy for restore object in the end of recursive function
+                */
+                const elementToSpecialize = this.elementsToSpecialize[0];
+                const copyElement = this.params[elementToSpecialize];
+
+                this.elementsToSpecialize.shift();
+                if (!(elementToSpecialize in this.params)) {
+                    const errMessage = 'No key of specialization found: ';
+                    throw new Error(errMessage + elementToSpecialize);
+                } else {
+                    copyElement.forEach(currentElement => {
+                        this.params[elementToSpecialize] = currentElement;
+                        this.execute();
+                    });
+                }
+                this.params[elementToSpecialize] = copyElement;
+                this.elementsToSpecialize.push(elementToSpecialize);
+            } else {
+                /**
+                * Create copy for not modify the current instance of the object
+                * during the call of the callback function
+                */
+                const matrixGenerated = new TestMatrix(this.params);
+                /**
+                * Search if the matrix we generated is an exception
+                */
+
+                const callFunction = (matrixFather, matrixChild, callback) => {
+                    const result = Object.keys(matrixChild.params)
+                    .every(currentKey => {
+                        return typeof matrixChild
+                        .params[currentKey] !== 'object';
+                    });
+
+                    if (result === true) {
+                        it(matrixChild.serialize(), (done) => {
+                            callback(matrixChild);
+                            done();
+                        });
+                    } else {
+                        describe(matrixChild.serialize(), () => {
+                            callback(matrixChild);
+                            matrixChild.execute();
+                        });
+                    }
+                };
+                let anExceptionWasFound = false;
+                this.listOfExceptions.forEach(exception => {
```

Comment:
I would have seen a filter approach here, so that it would look like so:
```
if (this.listOfExceptions.some(exception => isException(this, isException))) {
    this.listOfExceptions.filter(exception => isException(this, exception)).forEach(executeCallbacks);
} else {
   executeCallback();
}
```